source/mesh_refinement/compaction_length.cc

```diff
@@ -0,0 +1,155 @@
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/mesh_refinement/compaction_length.h>
+#include <aspect/melt.h>
+#include <aspect/simulator.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+    template <int dim>
+    void
+    CompactionLength<dim>::tag_additional_cells () const
+    {
+      // tag_additional_cells is executed before the equations are solved
+      // for the very first time. If we do not have the finite element, we
+      // do not have any material properties and just do nothing in this plugin.
+      if (this->get_dof_handler().n_locally_owned_dofs() == 0)
+        return;
+
+      const Quadrature<dim> quadrature(this->get_fe().base_element(this->introspection().base_elements.compositional_fields).get_unit_support_points());
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature,
+                               update_quadrature_points | update_values | update_gradients);
+
+      MaterialModel::MaterialModelInputs<dim> in(quadrature.size(), this->n_compositional_fields());
+      MaterialModel::MaterialModelOutputs<dim> out(quadrature.size(), this->n_compositional_fields());
+      MeltHandler<dim>::create_material_model_outputs(out);
+
+      const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+
+      for (typename DoFHandler<dim>::active_cell_iterator
+           cell = this->get_dof_handler().begin_active();
+           cell != this->get_dof_handler().end(); ++cell)
+        {
+          if (cell->is_locally_owned())
+            {
+              bool refine = false;
+              bool clear_coarsen = false;
+
+              fe_values.reinit(cell);
+              in.reinit(fe_values, cell, this->introspection(), this->get_solution(), true);
+              this->get_material_model().evaluate(in, out);
+
+              MaterialModel::MeltOutputs<dim> *melt_out = out.template get_additional_output<MaterialModel::MeltOutputs<dim> >();
+              AssertThrow(melt_out != NULL,
+                          ExcMessage("Need MeltOutputs from the material model for computing the melt properties."));
+
+              // for each temperature dof, check if the compaction length exceeds the cell size
+              for (unsigned int i=0; i<this->get_fe().base_element(this->introspection().base_elements.temperature).dofs_per_cell; ++i)
```

Comment:
you use the compositional field support points to initialize the quadrature, but you use the temperature support points here, is that on purpose? That looks incorrect.

Reply:
You're right, I copy/pasted it together from different places... I am using the composition everywhere now, which probably makes most sense.